### PR TITLE
Scala 3 support

### LINKED ISSUE #825: Scala 3 migration strategy
Hi, what is the plan to port the codebase to dotty? Is that a WIP somewhere, or is that scheduled sometime in the future?

I tried using it via withDottyCompat and ran into some issues. There were a bunch of typing translation issues that said something about existentials. Also it couldn't call Callback#runNow but I think it was a scala.js bug that may have been fixed already.

### PR DESCRIPTION
**Pull request by japgolly**
Closes #825

### CODE REVIEW

**Comment by rleibman** [comment]
So.... what's the scala3 support look like. Is it a good time to start porting my projects over?

**Comment by japgolly** [comment]
Oh hey @rleibman just saw your msg, I'd say it's ready for libraries to start porting over but I wouldn't switch an app over yet.

File: core/src/main/scala-3/japgolly/scalajs/react/component/builder/ComponentBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+
+  // ===================================================================================================================
+
+  /** You're on step 1/4 on the way to building a component.
+    *
+    * If your component is to be stateful, here you must specify it's initial state.
+    *
+    * If your component is to be stateless, you can skip this step or explicitly use `.stateless`.
+    */
+  final class Step1[P](private val name: String) extends AnyVal {
```
Comment (japgolly) [inline core/src/main/scala-3/japgolly/scalajs/react/component/builder/ComponentBuilder.scala:39]: try opaque type

File: core/src/main/scala-3/japgolly/scalajs/react/internal/Singleton.scala [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  implicit val Unit: Singleton[Unit] =
+    new Singleton(())(() => new js.Object)
+
+  implicit val BoxUnit: Singleton[Box[Unit]] =
+    new Singleton(Box.Unit)(() => Box(()))
+
+  // TODO: use erased class
+  sealed trait Not[A]
+  inline given noSingletonFor[A](using NotGiven[Singleton[A]]): Not[A] = null
```
Comment (japgolly) [inline core/src/main/scala-3/japgolly/scalajs/react/internal/Singleton.scala:22]: inline evidence

File: core/src/main/scala/japgolly/scalajs/react/CtorType.scala [hunk 3]
```diff
@@ -25,20 +25,26 @@ sealed abstract class CtorType[-P, +U] {
...
 
   final def withRawProp(name: String, value: js.Any): This[P, U] =
     addMod(_.asInstanceOf[js.Dynamic].updateDynamic(name)(value))
 
-  final def withKey(k: Key): This[P, U] =
-    withRawProp("key", k.asInstanceOf[js.Any])
-
-  final def withKey(k: Long): This[P, U] =
-    withKey(k.toString)
+  // TODO: https://github.com/lampepfl/dotty/issues/12265
```
Comment (japgolly) [inline core/src/main/scala/japgolly/scalajs/react/CtorType.scala:37]: retest this

File: core/src/main/scala/japgolly/scalajs/react/hooks/CustomHook.scala [hunk 4]
```diff
@@ -165,6 +170,7 @@ object CustomHook {
     CustomHook[() => D]
       .useState((0, Option.empty[D]))
       .buildReturning { (getDeps, depsState) =>
+
```
Comment (japgolly) [inline core/src/main/scala/japgolly/scalajs/react/hooks/CustomHook.scala:173]: no

File: core/src/main/scala/japgolly/scalajs/react/vdom/Exports.scala [hunk 5]
```diff
@@ -1,22 +1,10 @@
...
-
 object Exports extends Exports
-abstract class Exports {
 
-  private[vdom] implicit final def VdomExtString(a: String): ExportsStatic.VdomExtString =
-    new ExportsStatic.VdomExtString(a)
+abstract class Exports {
 
-  final val EmptyVdom: VdomNode            = japgolly.scalajs.react.vdom.VdomNode.empty
+  final val EmptyVdom: VdomNode             = japgolly.scalajs.react.vdom.VdomNode.empty
```
Comment (japgolly) [inline core/src/main/scala/japgolly/scalajs/react/vdom/Exports.scala:7]: formatting

File: extra/src/main/scala/japgolly/scalajs/react/extra/router/Urls.scala [hunk 6]
```diff
@@ -25,7 +25,9 @@ abstract class PathLike[Self <: PathLike[Self]] {
   * Examples: `BaseUrl("http://www.blah.com/hello")`,  `BaseUrl.fromWindowOrigin / "hello"`.
   */
 final case class BaseUrl(value: String) extends PathLike[BaseUrl] {
-  assertWarn(value contains "://", s"$this doesn't seem to be a valid URL. It's missing '://'. Consider using BaseUrl.fromWindowOrigin.")
+  devAssertWarn(
+    value contains "://",
+    s"$this doesn't seem to be a valid URL. It's missing '://'. Consider using BaseUrl.fromWindowOrigin.")
```
Comment (japgolly) [inline extra/src/main/scala/japgolly/scalajs/react/extra/router/Urls.scala:30]: confirm this is removed from prod

File: gh-pages/src/main/scala/ghpages/secret/tests/AsyncTest.scala [hunk 7]
```diff
@@ -28,7 +28,7 @@ object AsyncTest {
   }
 
   def testCmp2[A: Equal](body: AsyncCallback[A], expect1: A, prep2: () => Unit, expect2: A) = {
-    var complete  = null: Try[A] => Callback
+    var complete  = null: (Try[A] => Callback)
```
Comment (japgolly) [inline gh-pages/src/main/scala/ghpages/secret/tests/AsyncTest.scala:31]: report scala bug

File: test/src/main/scala/japgolly/scalajs/react/test/ReactTestUtils.scala [hunk 8]
```diff
@@ -19,8 +19,15 @@ object ReactTestUtils {
...
-  type MountedOutput = JsComponent.Mounted[_ <: js.Object, _ <: js.Object]
-  private def wrapMO(r: RawM | Null): MountedOutput = if (r == null) null else JsComponent.mounted(JsUtil.notNull(r))
+  type MountedOutput = JsComponent.Mounted[js.Object, js.Object]
+  private def wrapMO(r: RawM | Null): MountedOutput =
+    if (r == null)
+      null
+    else {
+      val r2 = JsUtil.notNull[RawM](r) // TODO: WTF??
+      val x = JsComponent.mounted(r2)
+      x.asInstanceOf[MountedOutput] // TODO: WTF??
```
Comment (japgolly) [inline test/src/main/scala/japgolly/scalajs/react/test/ReactTestUtils.scala:29]: raise issues

File: test/src/test/scala/japgolly/scalajs/react/CatsTest.scala [hunk 9]
```diff
@@ -22,19 +22,21 @@ object CatsTest extends TestSuite {
   val tests = Tests {
 
     "inference" - {
-      import japgolly.scalajs.react.test.InferenceUtil._
+      import japgolly.scalajs.react.test.InferenceHelpers._
+
+      assertType[Int].is[Int]
```
Comment (japgolly) [inline test/src/test/scala/japgolly/scalajs/react/CatsTest.scala:27]: remove

File: core/src/main/scala-3/japgolly/scalajs/react/component/builder/ComponentBuilder.scala (hunk 1 above)
Comment (japgolly) [reply core/src/main/scala-3/japgolly/scalajs/react/component/builder/ComponentBuilder.scala:39]: actually nah

**Pull request merged**